File: sparrowsong/20.go
package leetcode

func isValid(s string) bool {
	helper := map[byte]int{
		'(': -1, '{': -2, '[': -3, ')': 1, '}': 2, ']': 3,
	}
	stack := []int{}
	for _, b := range []byte(s) {
		i := helper[b]
		if i < 0 {
			stack = append(stack, i)
		} else {
			if len(stack) > 0 && -stack[len(stack)-1] == i {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		}
	}
	if len(stack) == 0 {
		return true
	}
	return false
}

File: LIUzq/20.cpp
class Solution {
public:
    bool isValid(string s) {
        if(s.empty()) return true;
        if(s.length()%2 != 0) return false;
       stack<char> a ; //char  not int  注意栈的声明怎么用
       for (auto ch: s){
            switch(ch){
                case '(':
                case'[':
                case'{':
                    a.push(ch);
                    continue;
                case')':
                    if(a.empty()|| a.top() != '(') //栈顶元素必须与前一个对应，不然就invalid
                    //top是返回栈顶元素，不删除，pop是删除栈顶元素
                    return false;
                    a.pop();
                    continue;
                case']':
                    if(a.empty() || a.top()!= '[')
                    return false;
                    a.pop();
                    continue;
                case'}':
                    if(a.empty()|| a.top() != '{')
                    return false;
                    a.pop();
                    continue;
            }
       }
       return a.empty();
    }
};